openwisp_controller/config/controller/views.py

```diff
@@ -52,62 +72,117 @@ class UpdateLastIpMixin(object):
+        # cause OpenWISP to be confused
+        if not device.management_ip:
+            return
+        queryset = self.model.objects.filter(
+            organization_id=device.organization_id, management_ip=device.management_ip
+        ).exclude(pk=device.pk)
+        for dupe in queryset.only('pk', 'key', 'management_ip'):
+            dupe.management_ip = ''
+            dupe.save(update_fields=['management_ip'])
+
+    def _remove_duplicated_last_ip(self, device):
+        # in the case of last_ip, we take a different approach,
+        # because it may be a public IP. If it's a public IP we will
+        # allow it to be duplicated
+        if not device.last_ip or not ip_address(device.last_ip).is_private:
+            return
+        queryset = Device.objects.filter(
+            organization_id=device.organization_id, last_ip=device.last_ip
+        ).exclude(pk=device.pk)
+        for dupe in queryset.only('pk', 'key', 'last_ip', 'management_ip'):
+            dupe.last_ip = ''
+            dupe.save(update_fields=['last_ip'])
+
 
-class ActiveOrgMixin(object):
+def get_device_args_rewrite(view):
     """
-    adds check to organization.is_active to ``get_object`` method
+    Use only the PK parameter for calculating the cache key
     """
-
-    def get_object(self, *args, **kwargs):
-        kwargs['organization__is_active'] = True
-        return super().get_object(*args, **kwargs)
+    # avoid ambiguity between hex format and dashed string
+    # return view.kwargs['pk']
+    pk = view.kwargs['pk']
+    try:
+        pk = uuid.UUID(pk)
+    except ValueError:
+        return pk
+    return pk.hex
 
 
-class DeviceChecksumView(ActiveOrgMixin, UpdateLastIpMixin, BaseConfigView):
+class DeviceChecksumView(UpdateLastIpMixin, GetDeviceView):
     """
     returns device's configuration checksum
     """
 
-    model = Device
-
-    def get(self, request, *args, **kwargs):
-        device = self.get_object(*args, **kwargs)
+    def get(self, request, pk):
+        device = self.get_device()
         bad_request = forbid_unallowed(request, 'GET', 'key', device.key)
         if bad_request:
             return bad_request
-        self.update_last_ip(device, request)
+        updated = self.update_last_ip(device, request)
+        # updates cache if ip addresses changed
+        if updated:
+            self.update_device_cache(device)
         checksum_requested.send(
             sender=device.__class__, instance=device, request=request
         )
-        return ControllerResponse(device.config.checksum, content_type='text/plain')
+        return ControllerResponse(
+            device.config.get_cached_checksum(), content_type='text/plain'
+        )
 
+    # while Config.get_cached_checksum() does not have a cache timeout
+    # in this case, since this view is reachable from the public internet,
+    # we avoid not setting a timeout to avoid
+    # the possibility of a Denial of Service attack
```

Comment:
I'm not sure 404 results get cached, I have to verify that.
If not, I can increase the timeout from 24 hours to something big like 10 days.